custom/userAuth.js

```diff
@@ -50,7 +52,34 @@ getAuth().then(({email, key}) => {
 
   // seralize/deseralization methods for extracting user information from the
   // session cookie and adding it to the req.passport object
-  passport.serializeUser((user, done) => done(null, user))
+
+  // serialize is called when the user first authenticates
+  // Use this callback as an opportunity to check their email against the chive permissions list
+  // and attach an "authorized" flag to their PassportJS object
+  passport.serializeUser(async (user, done) => {
+    user.authorized = false;
+
+    // magic Google Apps Script; takes a "user" parameter in a GET request and returns a JSON object with an "authorized" key
+    const authUrl = "https://script.google.com/macros/s/AKfycby3zj7uw2ZHi6JUcCt8aaRoE3Z8lUo3uKojxGXXIgeCoJ9h78CvXGJasUl--t23JfxB/exec"
+
+    try {
+      // TODO: Technically there could be more than one email attached to a profile - should we send all of them?
+      var authResp = await axios.get(authUrl, {params: {user: user.emails[0].value}})
```

Comment:
I'm not knowledgeable about the JS/Node ecosystem, but I'm curious why use a library here vs. the built-in `fetch` function. Just to avoid adding dependencies if we can.

Reply:
I can't remember what Node version we're on here (or where to check that) but Fetch wasn't always natively available in Node - it was a browser thing first. But if we do need a library, I would favor using one like `node-fetch` that directly imitates the Fetch API so it's easy to migrate later.

Reply:
Ah, we're on Node 18, where Fetch was still "experimental" - I'm not sure what limitations that implied.

One thing axios does give you that Fetch or a Fetch-imitating library doesn't is a more convenient API for JSON handling, but I don't think that's a very compelling advantage.

Reply:
(I was about to ask why not just use the native `https` module, but the answer to that would be that it doesn't return promises so you can't use `await`)

Reply:
Honestly, no reason in particular other than that googling how to do this (Node is not my expertise) brought up that it used to be requests, and now there are a bunch of options of which one is axios.

I didn't realize there was a good one built in, I can look into it.

Reply:
The built-in one is `http`/`https`. The disadvantage of that is that you have to use callbacks.

In Node 18, there is a built-in `fetch` that imitates the browser one, and supports await. It's marked as "experimental" but I believe it has all the functionality we would need, although it might cause messages to be logged about how you're using an experimental API. In Node 21+ they made it into a full official built-in.

_If_ for any reason the experimental one in Node 18 isn't quite good enough, that's why we might want to use the `node-fetch` package instead. If you add that, then you can do `const fetch = import('node-fetch')` at the top of the file and then use fetch in the rest of the code the same as you would for the built-in... so that if we're ever able to migrate to Node 21+, we can just remove that import and the code stays the same.

Reply:
Done. builtin fetch doesn't handle array params quite as cleanly, but the workaround is easy enough to justify not pulling in another module.